packages/zoe/src/contractSupport/recorder.js

```diff
@@ -0,0 +1,264 @@
+        this.state.closed = true;
+        return publisher.finish(value);
+      },
+    },
+  );
+
+  return makeRecorder;
+};
+harden(prepareRecorder);
+/** @typedef {ReturnType<typeof prepareRecorder>} MakeRecorder */
+
+/**
+ * @param {{makeRecorder: MakeRecorder, makeDurablePublishKit: ReturnType<typeof prepareDurablePublishKit>}} makers
+ */
+export const defineRecorderKit = ({ makeRecorder, makeDurablePublishKit }) => {
+  /**
+   * @template T
+   * @param {StorageNode} storageNode
+   * @param {TypedMatcher<T>} [valueGuard]
+   * @returns {RecorderKit<T>}
+   */
+  const makeRecorderKit = (storageNode, valueGuard) => {
+    const { subscriber, publisher } = makeDurablePublishKit();
+    const recorder = makeRecorder(publisher, storageNode, valueGuard);
+    return harden({ subscriber, recorder });
+  };
+  return makeRecorderKit;
+};
+/** @typedef {ReturnType<typeof defineRecorderKit>} MakeRecorderKit */
+
+/**
+ * @param {{makeRecorder: MakeRecorder, makeDurablePublishKit: ReturnType<typeof prepareDurablePublishKit>}} makers
+ */
+export const defineERecorderKit = ({ makeRecorder, makeDurablePublishKit }) => {
+  /**
+   * @template T
+   * @param {ERef<StorageNode>} storageNodeP
+   * @param {TypedMatcher<T>} [valueGuard]
+   * @returns {EventualRecorderKit<T>}
+   */
+  const makeERecorderKit = (storageNodeP, valueGuard) => {
+    const { publisher, subscriber } = makeDurablePublishKit();
+    const recorderP = E.when(storageNodeP, storageNode =>
+      makeRecorder(publisher, storageNode, valueGuard),
+    );
+    return { subscriber, recorderP };
+  };
+  return makeERecorderKit;
+};
+harden(defineERecorderKit);
+/** @typedef {ReturnType<typeof defineERecorderKit>} MakeERecorderKit */
+
+/**
+ * Convenience wrapper to prepare the DurablePublishKit and Recorder kinds.
+ * Note that because prepareRecorder() can only be called once per baggage,
+ * this should only be used when there is no need for an EventualRecorderKit.
+ * When there is, prepare the kinds separately and pass to the kit definers.
+ *
+ * @param {import('@agoric/vat-data').Baggage} baggage
+ * @param {ERef<Marshaller>} marshaller
+ */
+export const prepareRecorderKit = (baggage, marshaller) => {
+  const makeDurablePublishKit = prepareDurablePublishKit(
+    baggage,
+    'Durable Publish Kit',
+  );
+  const makeRecorder = prepareRecorder(baggage, marshaller);
+  return defineRecorderKit({ makeDurablePublishKit, makeRecorder });
+};
+
+/**
+ * To be called at most once per baggage.
+ *
+ * `makeRecorderKit` is suitable for making a durable `RecorderKit` which can be held in Exo state.
+ * `makeERecorderKit` is for closures that must return a `subscriber` synchronously but can defer the `recorder`.
+ *
+ * @param {import('@agoric/vat-data').Baggage} baggage
+ * @param {ERef<Marshaller>} marshaller
+ */
+export const prepareRecorderKitMakers = (baggage, marshaller) => {
+  const makeDurablePublishKit = prepareDurablePublishKit(
```

Comment:
It still feels overly obfuscated to have `prepareRecorderKitMakers` define two independent Kinds, and likely to bite us if a future change requires access to a non-Recorder publish kit. I really do think exposing and using the "define" pattern as the entry point is superior, but won't go further than a strong recommendation here.